cli/config.go

```diff
@@ -5,35 +5,95 @@ package cli
 
 import (
 	"fmt"
+	"io"
 	"log"
 	"os"
+	"reflect"
+	"strconv"
 
+	"github.com/mainflux/mainflux/internal/env"
+	mflog "github.com/mainflux/mainflux/logger"
 	"github.com/mainflux/mainflux/pkg/errors"
 	"github.com/pelletier/go-toml"
+	"github.com/spf13/cobra"
 )
 
+type remotes struct {
+	ThingsURL      string `toml:"things_url"`
+	UsersURL       string `toml:"users_url"`
+	ReaderURL      string `toml:"reader_url"`
+	HTTPAdapterURL string `toml:"http_adapter_url"`
+	BootstrapURL   string `toml:"bootstrap_url"`
+	CertsURL       string `toml:"certs_url"`
+}
+
+type filter struct {
+	Offset          string `toml:"offset"`
+	Limit           string `toml:"limit"`
+	RawOutput       string `toml:"raw_output"`
+	Name            string `toml:"name"`
+	Contact         string `toml:"contact"`
+	Email           string `toml:"email"`
+	Metadata        string `toml:"metadata"`
+	MsgContentType  string `toml:"msg_content_type"`
+	TLSVerification string `toml:"tls_verification"`
+}
+
+type channel struct {
+	Status string `toml:"status"`
+	State  string `toml:"state"`
+	Topic  string `toml:"topic"`
+}
+
 type Config struct {
-	Offset    uint   `toml:"offset"`
-	Limit     uint   `toml:"limit"`
-	Name      string `toml:"name"`
-	RawOutput bool   `toml:"raw_output"`
+	Remotes remotes `toml:"remotes"`
+	Filter  filter  `toml:"filter"`
+	Channel channel `toml:"channel"`
+}
+
+type config struct {
+	LogLevel string `env:"MF_CLI_LOG_LEVEL"   envDefault:"info"`
 }
 
 // read - retrieve config from a file.
 func read(file string) (Config, error) {
-	data, err := os.ReadFile(file)
 	c := Config{}
+	data, err := os.Open(file)
+	if err != nil {
+		return c, errors.Wrap(errors.New("Failed to read config file."), err)
+	}
+	defer data.Close()
+
+	cfg := config{}
+	if err := env.Parse(&cfg); err != nil {
+		log.Fatalf("failed to load configuration")
```

Comment:
Don't use fatal after a defer statement.
 Also keep fatal in the main function only. In case of other functions simply log an error and return to allow deferred actions to complete.

Reply:
Return an error you can check for in main()

Reply:
I have used logError defined in cli/utils